### PR TITLE
Week five topic slides

### PR DESCRIPTION
**Pull request by kberzinch**
The Trello card on this was really brief - I fleshed out what I could think of but this still seems really short.

Advice on expanding this deck somewhat is appreciated.

### CODE REVIEW

**Review by JasonGibson274 — COMMENTED** [review]
I would have some examples of polymorphism, inheritance and how/why we group methods and variables together in different classes.

File: slides/week5/week5.org [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+#+BEGIN_SRC c
+class example {
+    private:         // private is also implied if you don't specify visibility
+        int a;
+    public:
+        int getA() { return a; }
+};
+#+END_SRC
+
+* Why bother?
```
Comment (jgkamat) [inline slides/week5/week5.org:30]: Maybe you could discuss code reuse a bit here

File: slides/week5/week5.org [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+* What are classes?
+- Definitions for objects, i.e. what data and methods they have
+- An object is called an *instance* of a class
+
+* A simple example
+#+BEGIN_NOTES
+A class can have some data and some methods, which can either be public or private. Private means only visible to the class itself, public means visible to any other file. Drop to a terminal and use/modify this class.
+#+END_NOTES
+#+BEGIN_SRC c
```
Comment (jgkamat) [inline slides/week5/week5.org]: These should be `C++` not `c` (they produce compiler errors, and don't recognize C++ only keywords in highlighting).

File: slides/week5/week5.org [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+* Inheritance
+- Classes can inherit members from other classes
+#+BEGIN_SRC c
+class child: public example {
+    // we get `a` and `getA()` from example
+    private:
+        int b;
+    public:
+        int getB() { return b; }
```
Comment (jgkamat) [inline slides/week5/week5.org:60]: maybe show a `getAPlusB` method to show how it actually is working

File: slides/week5/week5.org [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+};
+#+END_SRC
+
+* Polymorphism
+- Now we can create =child= objects with all the properties of an =example= object
+- This means we can safely cast a =child= object to an =example= object
+- TODO I'm not a fan of this slide layout - ideas?
+#+BEGIN_SRC c
```
Comment (jgkamat) [inline slides/week5/week5.org]: Maybe:
```
- Now we can create =child= objects with all the properties of an =example= object
- This means we can safely cast a =child= object to an =example= object
  #+BEGIN_SRC C++
    child c;
    example& e = dynamic_cast<example&>(c);
  #+END_SRC
- But we can't go the other way
  #+BEGIN_SRC C++
    example e;
    child& c = dynamic_cast<child&>(e);
  #+END_SRC

  #+BEGIN_EXAMPLE
    example.cpp: In function ‘int main()’:
    example.cpp:20:38: error: cannot dynamic_cast ‘e’ (of type ‘class example’) to type
    ‘class child&’  (source type is not polymorphic)
         child& c = dynamic_cast<child&>(e);
  #+END_EXAMPLE
```

**Review by jgkamat — CHANGES_REQUESTED** [review]
Overall looking pretty good, but needs a few changes.

File: slides/week5/week5.org (hunk 3 above)
Comment (kberzinch) [reply slides/week5/week5.org:60]: Ooh, I like that

File: slides/week5/week5.org (hunk 4 above)
Comment (kberzinch) [reply slides/week5/week5.org]: I tried that but the bullet point in the middle didn't line up with the rest of them and it looked kind of dumb

Comment (jgkamat) [reply slides/week5/week5.org]: Hmm, that's weird. This is what I get, fwiw.

2017-07-31-212722_1920x1080_scrot

**Comment by ebarnette** [comment]
@kberzinch When you talk about what a class is, one of the best metaphors to use is that classes are like instructions/blueprints sent to a factory to build a lot of the same thing. "Definitions for objects" is a little vague. I feel like you can flesh out that slide more.

File: slides/week5/week5.org (hunk 3 above)
Comment (ebarnette) [reply slides/week5/week5.org:60]: Major plus 1 on this. The slide on inheritance is hella vague.

File: slides/week5/week5.org (hunk 1 above)
Comment (ebarnette) [reply slides/week5/week5.org:30]: +1

File: slides/week5/week5.org (hunk 3 above)
Comment (kberzinch) [reply slides/week5/week5.org:60]: Added

File: slides/week5/week5.org (hunk 1 above)
Comment (kberzinch) [reply slides/week5/week5.org:30]: I added a bullet point but I'm not a fan of the phrasing. Suggestions appreciated.

File: slides/week5/week5.org (hunk 4 above)
Comment (kberzinch) [reply slides/week5/week5.org]: OHHH I see what you did there. Fixed and pushed!

File: slides/week5/week5.org (hunk 1 above)
Comment (jgkamat) [reply slides/week5/week5.org:30]: Maybe: "Additional functionality can be added with minimal extra code using inheritance"?

**Review by jgkamat — CHANGES_REQUESTED** [review]
The content here looks mostly good to me, but there needs to be more content explaining what an object and what a class is (and what's the difference). I think the existing example jumps into code too early.

File: slides/week5/week5.org (hunk 1 above)
Comment (kberzinch) [reply slides/week5/week5.org:30]: I like it - added

**Review by jgkamat — DISMISSED** [review]
Overall, I think that the content of the slides looks good to me, but it still feels confusing for a newcomer. This topic is really hard though, so we should just focus on answering questions imo.

File: slides/week5/week5.org [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+* What are we doing today?
+- Classes in C++
+- Handling merge conflicts in =git=
+
+* What are classes?
+- Definitions for objects, i.e. what data and methods they have
+ - In other words, a blueprint for how to create an object
+- An object is called an *instance* of a class
+ - An object has all the functions and data structures defined in its class
+ - It doesn't share any data with other objects
```
Comment (JasonGibson274) [inline slides/week5/week5.org]: I feel like this is misleading. Objects can share data with other objects. Perhaps something like "Objects of the same class share the same schema of data but the actual values stored are independent instance to instance"

**Review by JasonGibson274 — CHANGES_REQUESTED** [review]
Overall looks good to me.

**Comment by JasonGibson274** [comment]
Actually we are missing the virtual keyword explanation.

**Comment by kberzinch** [comment]
Added @JasonGibson274

**Review by JasonGibson274 — DISMISSED** [review]

**Review by jgkamat — DISMISSED** [review]

File: slides/week5/week5.org [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+- Software projects are easier to understand when related data and functions are grouped together
+- Abstracting data behind getters/setters allows you to validate inputs from other parts of your application
+- Abstracting complex tasks into class methods enables simpler, easier to read high-level code
+- Classes can be extended to add functionality with minimal extra code using inheritance
+
+* Static members
+#+BEGIN_NOTES
+ Drop to a terminal and use this class.
+#+END_NOTES
+- For when you want data or functions to be part of your class, but they don't need to be "attached" to an instance of the class
```
Comment (joshhting) [inline slides/week5/week5.org:40]: This wording seems a little more confusing than explicitly saying that all instances of a class read from and write to the same static variable

Comment (kberzinch) [reply slides/week5/week5.org:40]: Added.

**Review by joshhting — APPROVED** [review]

**Pull request merged**